lib/ex_doc/formatter/html/templates.ex

```diff
@@ -146,6 +153,31 @@ defmodule ExDoc.Formatter.HTML.Templates do
     %{key: HTML.text_to_id(group), name: group, nodes: nodes}
   end
 
+  defp module_sections(module) do
+    sections =
+      case module.rendered_doc do
+        nil ->
+          []
+
+        moduledoc ->
+          moduledoc
+          |> extract_headers()
+          |> Enum.map_reduce(%{}, fn header, acc ->
+            case Map.fetch(acc, header.id) do
+              {:ok, id} ->
+                {%{header | anchor: "module-#{header.anchor}-#{id}"},
+                 Map.put(acc, header.id, id + 1)}
+
+              :error ->
+                {%{header | anchor: "module-#{header.anchor}"}, Map.put(acc, header.id, 1)}
+            end
+          end)
+          |> elem(0)
```

Comment:
Do we have this logic elsewhere? If so, we can re-use it? Or would it be too complicated?

Reply:
We do

```elixir
  @heading_regex ~r/<(h[23]).*?>(.*?)<\/\1>/m
  @spec link_headings(String.t(), Regex.t(), String.t()) :: String.t()
  def link_headings(content, regex \\ @heading_regex, prefix \\ "")
  def link_headings(nil, _, _), do: nil

  def link_headings(content, regex, prefix) do
    regex
    |> Regex.scan(content)
    |> Enum.reduce({content, %{}}, fn [match, tag, title], {content, occurrences} ->
      possible_id = HTML.text_to_id(title)
      id_occurred = Map.get(occurrences, possible_id, 0)

      anchor_id = if id_occurred >= 1, do: "#{possible_id}-#{id_occurred}", else: possible_id
      replacement = link_heading(match, tag, title, anchor_id, prefix)
      linked_content = String.replace(content, match, replacement, global: false)
      incremented_occs = Map.put(occurrences, possible_id, id_occurred + 1)
      {linked_content, incremented_occs}
    end)
    |> elem(0)
  end
```

The problem I see is that this is working on the derivative rendered data, and not the AST. A lot of this code could be greatly simplified if these actions were taken upstream of the rendering part of the process

Reply:
To answer your question. I suppose I could consolidate this logic, but its a much bigger project. I don't feel great about duplicating it.

If its any reconciliation, I would like to help to refactor this code to make it easier to work with.

Reply:
I understand. Yes, we are slowly migrating to the AST but I don’t think we did so for extras pages, so perhaps that’s the first step before we can address the duplication.